sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -409,6 +505,340 @@ class GpuDynamicPartitionDataWriter(
+ *    partition data if the total caches exceed the limitation.
+ *  - Step 2: If number of concurrent writers exceeds limit, sort rest of rows on partition.
+ *    Write rows one by one, and eagerly close the writer when finishing
+ *  Caller is expected to call `writeWithIterator()` instead of `write()` to write records.
+ */
+class GpuDynamicPartitionDataConcurrentWriter(
+    description: GpuWriteJobDescription,
+    taskAttemptContext: TaskAttemptContext,
+    committer: FileCommitProtocol,
+    spec: GpuConcurrentOutputWriterSpec)
+    extends GpuDynamicPartitionDataSingleWriter(description, taskAttemptContext, committer)
+        with Arm with Logging {
+
+  // Keep all the unclosed writers, key is partition directory string.
+  // Note: if fall back to sort-based mode, also use the opened writers in the map.
+  private val concurrentWriters = mutable.HashMap[String, WriterStatusWithCaches]()
+
+  // guarantee to close the caches and writers when task is finished
+  TaskContext.get().addTaskCompletionListener[Unit](_ => closeCachesAndWriters())
+
+  private val outDataTypes = description.dataColumns.map(_.dataType).toArray
+
+  override def abort(): Unit = {
+    try {
+      closeCachesAndWriters()
+    } finally {
+      committer.abortTask(taskAttemptContext)
+    }
+  }
+
+  /** Wrapper class for status of a unique concurrent output writer. */
+  private class WriterStatusWithCaches(
+      // writer status
+      var writerStatus: WriterStatus,
+
+      // caches for this partition or writer
+      val tableCaches: ListBuffer[Table] = ListBuffer(),
+
+      // current device bytes for the above caches
+      var deviceBytes: Long = 0
+  )
+
+  /**
+   * State to indicate if we are falling back to sort-based writer.
+   * Because we first try to use concurrent writers, its initial value is false.
+   */
+  private var fallBackToSortBased: Boolean = false
+
+  def writeWithSingleWriter(cb: ColumnarBatch): Unit = {
+    // invoke `GpuDynamicPartitionDataSingleWriter`.write
+    super.write(cb, partDir => concurrentWriters.get(partDir).map(e => e.writerStatus))
+  }
+
+  def writeWithConcurrentWriter(cb: ColumnarBatch): Unit = {
+    this.write(cb)
+  }
+
+  /**
+   *  Write an iterator of column batch.
+    * @param iterator the iterator of column batch
+   */
+  override def writeWithIterator(iterator: Iterator[ColumnarBatch]): Unit = {
+    // 1: try concurrent writer
+    while (iterator.hasNext && !fallBackToSortBased) {
+      if(concurrentWriters.size >= spec.maxWriters) {
+        // mark falling back
+        fallBackToSortBased = true
+      } else {
+        // concurrently write and update the `concurrentWriters` map
+        writeWithConcurrentWriter(iterator.next())
+      }
+    }
+
+    // 2: will fall back to single writer, first flush all the caches in the concurrent writer step.
+    concurrentWriters.filter(pair => pair._2.tableCaches.nonEmpty)
```

Comment:
I personally would much rather have us not flush the caches, especially once we make the linked list spillable. That way we don't get tiny row groups when this happens.

Reply:
Done